text/0008-kadalu-kubernetes-storage-monitoring.adoc

```diff
@@ -0,0 +1,153 @@
+
+.Prometheus Metrics
+----
+GET /metrics
+----
+
+.JSON output
+----
+GET /metrics.json
+----
+
+=== Types of Metrics
+
+==== Count Metrics
+
+Metrics related to the resource counts.
+
+- Number of Storages
+- Number of PVs
+- Number of Pods in the Kadalu namespace.
+
+==== Health Metrics
+
+Metrics related to the state of the resources.
+
+- How to make sure that the Kadalu Operator is running successfully?
+- How to know if the CSI Provisioner stopped processing further requests?
+- How to know if the Self heal daemon is stopped working?
+- How to know if a resource is restarted?
+
+==== Utilization Metrics
+
+Metrics related to the resource utilization like Storage, CPU, Memory, Inodes etc.
+
+- How to know which Storage pool is getting full?
+- How to know if any resource is consuming more CPU or Memory?
+
+==== Performance Metrics
+
+Metrics related to Storage I/O performance, Provisioning, mount performance.
+
+- How to measure if the Storage Performance is improved or degraded?
+- How quick a PV is provisioned?
+
+==== Events
+
+A few Events can be derived from the collected metrics by comparing with the latest data with the previously collected metrics. For example,
+
+- *Restarted Event* - If the current uptime of a resource is less than the previous one.
+- *Upgraded Event* - If the current version is greater than the previous one.
+- *Option Change* - If the list of Volume options is different than the previously collected options data.
+- *New Storage added* - When the latest list of metrics shows a new Storage compared to previously collected data.
+
+These events will be exposed by `GET /events` URL. These events also helps to understand the storage better when viewed along with the historical metrics. For example, if a Performance degradation is observed in the chart and associated event says that a Option is changed or one Replica server Pod was down or the Cluster is upgraded to a new Version.
+
+==== Alerts
+
+All events may not be useful, some events may be very noisy. For example, an Event related to changes to utilization is very noisy. But becomes very important when it meets some criteria.
+
+- When utilization crosses 70%.
+- Performance improvement/degrade.
+- Storage is down.
+
+=== List of Metrics
+
+- *Number of Storage pools*
+- *Number of PVs*
+- *Number of Storage Units/Bricks*
+- *Operator Health* - Operator is running or not
+- *Storage Units Health* - Health of Brick processes, `1 for Up, and 0 for Down`
+- *Self Heal daemon Health*
+- *Health of Metrics exporter*
+- *CSI Provisioner Health*
+- *CSI/Quotad health*
+- *CSI/Mounts health* (Based on expected number of Volumes in ConfigMap and number of mount processes). Gluster client process will continue to run even if all the bricks are down, it waits for the brick processes and re-connects as soon as they are available. Detect this by doing a regular IO from the mount or parsing the log files for `ENOTCONN` errors.
```

Comment:
> "regular IO from the mount"
- Please clarify which mount will be used for performing this op, the provisioner with some test dir or a new pod etc?

Reply:
From the mount available in the CSI provisioner pod.